apis/v1alpha2/clusternetworkpolicy_types.go

```diff
@@ -0,0 +1,436 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields in this package are required unless Explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:shortName=cnp,scope=Cluster
+// +kubebuilder:printcolumn:name="Tier",type=string,JSONPath=".spec.tier"
+// +kubebuilder:printcolumn:name="Priority",type=string,JSONPath=".spec.priority"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// ClusterNetworkPolicy is a cluster-wide network policy resource.
+type ClusterNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Specification of the desired behavior of ClusterNetworkPolicy.
+	Spec ClusterNetworkPolicySpec `json:"spec"`
+
+	// Status is the status to be reported by the implementation.
+	// +optional
+	Status ClusterNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// ClusterNetworkPolicyStatus defines the observed state of ClusterNetworkPolicy.
+type ClusterNetworkPolicyStatus struct {
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// ClusterNetworkPolicySpec defines the desired state of ClusterNetworkPolicy.
+type ClusterNetworkPolicySpec struct {
+	// Tier is used as a first layer of rule prioritization.
+	// ClusterNetworkPolicy has 2 tiers: admin and baseline.
+	// admin tier is used to set strict security rules that can't be overridden.
+	// baseline tier is used to set the default network policy different from "allow all", that can be overridden by NetworkPolicy.
+	// While ClusterNetworkPolicy can be created in the admin and baseline tiers, other types of policy occupy their own tiers:
+	// NetworkPolicy tier contains all NetworkPolicies in the cluster and
+	// DefaultTier, which is the default cluster policy, allows all pods to communicate with each other.
+	//
+	// Tiers are evaluated top to bottom, in the following order:
+	// 1. Admin tier
+	// 2. NetworkPolicy tier
+	// 3. Baseline tier
+	// 4. Default tier
+	Tier Tier `json:"tier,omitempty"`
```

Comment:
I really don't understand this very well, so the enum only has two tiers by there can be two additional more?

how does it work?

Reply:
yeah we may need to figure out a better way to explain this. Some tiers (Admin and Baseline) belong to the ClusterNetworkPolicy, but there are other tiers in the cluster (NetworkPolicy and Default).
It only defines in which order the policies apply, so in that sense nothing has changed: ANP was always applied before NetworkPolicy and BANP was always applied after NetworkPolicy.
The "new" term here is the Default tier, which simply implies the default "allow-all" policy of any k8s cluster, which just means that without any policies all traffic is allowed.
lmk if you have more questions :)

Reply:
that is good feedback, let us think about other options for the Default tier name